src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -0,0 +1,261 @@
+﻿using Npgsql.Util;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Npgsql
+{
+    sealed class MultiHostConnectorPool : ConnectorPoolBase
+    {
+        readonly ConnectorPool[] _pools;
+
+        internal override (int Total, int Idle, int Busy) Statistics
+        {
+            get
+            {
+                var numConnectors = 0;
+                var idleCount = 0;
+
+                foreach (var pool in _pools)
+                {
+                    numConnectors += pool.NumConnectors;
+                    idleCount += pool.IdleCount;
+                }
+                
+                return (numConnectors, idleCount, numConnectors - idleCount);
+            }
+        }
+
+        public MultiHostConnectorPool(NpgsqlConnectionStringBuilder settings, string connString) : base(settings, connString)
+        {
+            var hosts = settings.Host!.Split(',');
+            _pools = new ConnectorPool[hosts.Length];
+            for (var i = 0; i < hosts.Length; i++)
+            {
+                var host = hosts[i];
+                var port = settings.Port;
+                // TODO: Add support for host:port to the default pool (and conn without pooling)
+                var portSeparator = host.IndexOf(':');
+                if (portSeparator != -1)
+                {
+                    host = host.Substring(0, portSeparator);
+                    port = int.Parse(host.Substring(portSeparator + 1));
+                }
+                var poolSettings = settings.Clone();
+                poolSettings.Host = host;
+                poolSettings.Port = port;
+                _pools[i] = new ConnectorPool(poolSettings, poolSettings.ConnectionString, this);
+                // TODO: should be safely removed
+                /*
+                // Setting the default settings due to PoolManager.TryGetValue
+                _pools[i] = new ConnectorPool(poolSettings, settings.ConnectionString, this);
+                */
+            }
+        }
+
+        static bool IsValidHost(ConnectorPool pool, TargetServerType preferedType)
+        {
+            var currentState = GetHostState(pool);
+            if (currentState == HostState.Offline)
+                return false;
+            return preferedType == TargetServerType.Any || currentState == HostState.Unknown ||
+                preferedType.HasFlag(TargetServerType.Primary) && currentState == HostState.Primary ||
+                preferedType.HasFlag(TargetServerType.Secondary) && currentState == HostState.Secondary;
+        }
+
+        static bool IsValidUnpreferedHost(ConnectorPool pool, TargetServerType preferedType)
```

Comment:
Shouldn't this also accept preferred pools, i.e. return true for any pool which would pass IsValidHost above? This is for the case where a race condition changes the state in between.

Reply:
The idea here was that:
1) on the first 'cycle' we attempt to return the preferred host
2) on the second 'cycle' we attempt to return the non-preferred host

I guess we can also check for the preferred host again on the second 'cycle'.